File: scripts/mcp_compat/tianyancha_stdio_bridge.mjs
#!/usr/bin/env node

import fs from "node:fs";
import os from "node:os";
import path from "node:path";

function loadOpenClawEnv() {
  try {
    const cfg = JSON.parse(
      fs.readFileSync(path.join(os.homedir(), ".openclaw", "openclaw.json"), "utf8"),
    );
    return cfg?.env || {};
  } catch {
    return {};
  }
}

const OPENCLAW_ENV = loadOpenClawEnv();
const REMOTE_URL = OPENCLAW_ENV.TIANYANCHA_MCP_URL;
const AUTHORIZATION = OPENCLAW_ENV.TIANYANCHA_AUTHORIZATION;

let sessionId = null;
let readBuffer = Buffer.alloc(0);

const TOOLS = [
  {
    name: "companyBaseInfo",
    description: "查询企业工商登记与基础信息，用于确认企业主体、法人、注册资本、成立时间与经营状态。",
    inputSchema: {
      type: "object",
      properties: {
        companyName: {
          type: "string",
          description: "企业精确名称",
        },
      },
      required: ["companyName"],
      additionalProperties: true,
    },
  },
  {
    name: "risk",
    description: "查询企业法律、经营与风险信号，用于客户准入前的风险筛查。",
    inputSchema: {
      type: "object",
      properties: {
        companyName: {
          type: "string",
          description: "企业精确名称",
        },
      },
      required: ["companyName"],
      additionalProperties: true,
    },
  },
];

function log(message) {
  void message;
}

function writeMessage(message) {
  const payload = Buffer.from(JSON.stringify(message), "utf8");
  log(`OUT ${payload.toString("utf8")}`);
  process.stdout.write(`Content-Length: ${payload.length}\r\n\r\n`);
  process.stdout.write(payload);
}

function sendError(id, message) {
  writeMessage({ jsonrpc: "2.0", id, error: { code: -32000, message } });
}

function sendResponse(id, result) {
  writeMessage({ jsonrpc: "2.0", id, result });
}

function tryParseMessage() {
  const delimiter = readBuffer.indexOf("\r\n\r\n");
  if (delimiter === -1) return null;
  const headerText = readBuffer.subarray(0, delimiter).toString("utf8");
  const headers = Object.fromEntries(
    headerText
      .split("\r\n")
      .map((line) => line.split(/:\s*/, 2))
      .filter(([key, value]) => key && value),
  );
  const length = Number.parseInt(headers["Content-Length"] || headers["content-length"] || "0", 10);
  if (!Number.isFinite(length) || length <= 0) {
    readBuffer = readBuffer.subarray(delimiter + 4);
    return null;
  }
  const bodyStart = delimiter + 4;
  const bodyEnd = bodyStart + length;
  if (readBuffer.length < bodyEnd) return null;
  const body = readBuffer.subarray(bodyStart, bodyEnd).toString("utf8");
  readBuffer = readBuffer.subarray(bodyEnd);
  log(`IN ${body}`);
  return JSON.parse(body);
}

function parseSseBody(body) {
  const chunks = [];
  for (const block of body.split("\n\n")) {
    for (const line of block.split(/\r?\n/)) {
      if (line.startsWith("data:")) chunks.push(line.slice(5).trim());
    }
  }
  if (!chunks.length) return null;
  return JSON.parse(chunks.join(""));
}

async function remotePost(payload, allowMissingSession = false) {
  if (!REMOTE_URL || !AUTHORIZATION) {
    throw new Error("Tianyancha environment is not configured");
  }
  const headers = {
    Authorization: AUTHORIZATION,
    "Content-Type": "application/json",
    Accept: "application/json, text/event-stream",
  };
  if (sessionId) headers["mcp-session-id"] = sessionId;
  const response = await fetch(REMOTE_URL, {
    method: "POST",
    headers,
    body: JSON.stringify(payload),
  });
  const nextSessionId = response.headers.get("mcp-session-id");
  if (nextSessionId) sessionId = nextSessionId;
  const text = await response.text();
  if (!response.ok && !(allowMissingSession && response.status === 400)) {
    throw new Error(`Tianyancha request failed (${response.status}): ${text.slice(0, 400)}`);
  }
  const contentType = response.headers.get("content-type") || "";
  if (!text.trim()) return null;
  if (contentType.includes("text/event-stream")) return parseSseBody(text);
  return JSON.parse(text);
}

function normalizeToolArgs(toolName, toolArgs = {}) {
  if ((toolName === "companyBaseInfo" || toolName === "risk") && toolArgs.companyName && !toolArgs.keyword) {
    return { ...toolArgs, keyword: toolArgs.companyName };
  }
  return toolArgs;
}

async function callTool(toolName, toolArgs = {}) {
  const response = await remotePost(
    {
      jsonrpc: "2.0",
      id: "cli-call",
      method: "tools/call",
      params: { name: toolName, arguments: normalizeToolArgs(toolName, toolArgs) },
    },
    true,
  );
  if (response?.error) {
    throw new Error(response.error.message || JSON.stringify(response.error));
  }
  const text = response?.result?.content?.map((item) => item?.text || "").join("\n").trim();
  if (!text) return response?.result ?? response ?? {};
  try {
    return JSON.parse(text);
  } catch {
    return text;
  }
}

async function handleMessage(message) {
  const { id, method } = message;
  if (method === "initialize") {
    sendResponse(id, {
      protocolVersion: message.params?.protocolVersion || "2025-03-26",
      capabilities: {
        tools: { listChanged: false },
        resources: { listChanged: false },
      },
      serverInfo: { name: "tianyancha-codex-bridge", version: "1.1.0" },
    });
    return;
  }
  if (method === "notifications/initialized") {
    try {
      await remotePost({ jsonrpc: "2.0", method: "notifications/initialized", params: {} }, true);
    } catch {
      // no-op
    }
    return;
  }
  if (method === "tools/list") {
    sendResponse(id, { tools: TOOLS });
    return;
  }
  if (method === "resources/list") {
    sendResponse(id, { resources: [] });
    return;
  }
  if (method === "resources/templates/list") {
    sendResponse(id, { resourceTemplates: [] });
    return;
  }
  const response = await remotePost(message, true);
  if (response) {
    writeMessage(response);
  } else if (id !== undefined) {
    sendResponse(id, {});
  }
}

async function runCli() {
  const [, , toolName, rawArgs = "{}"] = process.argv;
  if (!toolName) return false;
  try {
    const parsedArgs = JSON.parse(rawArgs);
    const data = await callTool(toolName, parsedArgs);
    process.stdout.write(`${JSON.stringify(data, null, 2)}\n`);
    return true;
  } catch (error) {
    process.stderr.write(`${error instanceof Error ? error.message : String(error)}\n`);
    process.exit(1);
  }
}

if (await runCli()) {
  process.exit(0);
}

process.stdin.on("data", async (chunk) => {
  readBuffer = Buffer.concat([readBuffer, chunk]);
  while (true) {
    const message = tryParseMessage();
    if (!message) break;
    try {
      await handleMessage(message);
    } catch (error) {
      if (message.id !== undefined) {
        sendError(message.id, `Tianyancha bridge error: ${error instanceof Error ? error.message : String(error)}`);
      }
    }
  }
});

process.stdin.on("end", () => process.exit(0));
process.on("uncaughtException", (error) => {
  log(`UNCAUGHT ${error?.stack || error}`);
  process.exit(1);
});
process.on("unhandledRejection", (error) => {
  log(`UNHANDLED ${error?.stack || error}`);
  process.exit(1);
});

File: scripts/mcp_compat/tianyancha_stdio_bridge.py
#!/usr/bin/env python3
import json
import os
import sys
import urllib.request
from pathlib import Path
from typing import Any, Dict, Optional


def iter_openclaw_config_paths() -> list[Path]:
    home = Path.home()
    preferred = home / ".openclaw" / "openclaw.json"
    others = sorted(home.glob(".openclaw*/openclaw.json"))
    paths: list[Path] = []
    if preferred.exists():
        paths.append(preferred)
    for candidate in others:
        if candidate not in paths:
            paths.append(candidate)
    return paths


def load_openclaw_config() -> Dict[str, Any]:
    for config_path in iter_openclaw_config_paths():
        try:
            return json.loads(config_path.read_text(encoding="utf-8"))
        except Exception:
            continue
    return {}


def load_best_provider_config(provider_id: str) -> Dict[str, Any]:
    for config_path in iter_openclaw_config_paths():
        try:
            config = json.loads(config_path.read_text(encoding="utf-8"))
        except Exception:
            continue
        providers = (((config.get("models") or {}).get("providers")) or {})
        provider = providers.get(provider_id)
        if isinstance(provider, dict) and provider:
            return config
    return load_openclaw_config()


def resolve_tianyancha_settings(config: Dict[str, Any]) -> Dict[str, str]:
    env_url = os.environ.get("TIANYANCHA_URL") or os.environ.get("TIANYANCHA_MCP_URL")
    env_auth = os.environ.get("TIANYANCHA_AUTHORIZATION")
    if env_url and env_auth:
        return {"url": env_url, "authorization": env_auth}
    providers = (((config.get("models") or {}).get("providers")) or {})
    provider = providers.get("tianyancha") or {}
    url = provider.get("baseUrl") if isinstance(provider.get("baseUrl"), str) else ""
    auth = provider.get("apiKey") if isinstance(provider.get("apiKey"), str) else ""
    if auth == "TIANYANCHA_AUTHORIZATION":
        auth = os.environ.get("TIANYANCHA_AUTHORIZATION", "")
    return {"url": url, "authorization": auth}


CONFIG = load_best_provider_config("tianyancha")
SETTINGS = resolve_tianyancha_settings(CONFIG)
REMOTE_URL = SETTINGS["url"]
AUTHORIZATION = SETTINGS["authorization"]
SESSION_ID: Optional[str] = None
INITIALIZED = False
LOG_PATH = os.environ.get("CODEX_MCP_DEBUG_LOG", "/tmp/tianyancha_codex_bridge.log")


def write_message(message: Dict[str, Any]) -> None:
    log(f"OUT {json.dumps(message, ensure_ascii=False)}")
    payload = json.dumps(message, ensure_ascii=False).encode("utf-8")
    sys.stdout.buffer.write(f"Content-Length: {len(payload)}\r\n\r\n".encode("ascii"))
    sys.stdout.buffer.write(payload)
    sys.stdout.buffer.flush()


def read_message() -> Optional[Dict[str, Any]]:
    headers: Dict[str, str] = {}
    while True:
        line = sys.stdin.buffer.readline()
        if not line:
            return None
        if line in (b"\r\n", b"\n"):
            break
        decoded = line.decode("utf-8").strip()
        if ":" in decoded:
            key, value = decoded.split(":", 1)
            headers[key.lower()] = value.strip()
    content_length = int(headers.get("content-length", "0"))
    if content_length <= 0:
        return None
    body = sys.stdin.buffer.read(content_length)
    if not body:
        return None
    message = json.loads(body.decode("utf-8"))
    log(f"IN {json.dumps(message, ensure_ascii=False)}")
    return message


def log(message: str) -> None:
    try:
        with open(LOG_PATH, "a", encoding="utf-8") as fh:
            fh.write(message + "\n")
    except Exception:
        pass


def parse_sse_body(body: str) -> Optional[Dict[str, Any]]:
    chunks = []
    for block in body.split("\n\n"):
        for line in block.splitlines():
            if line.startswith("data:"):
                chunks.append(line[5:].strip())
    if not chunks:
        return None
    return json.loads("".join(chunks))


def remote_post(payload: Dict[str, Any], session_id: Optional[str]) -> Optional[Dict[str, Any]]:
    headers = {
        "Authorization": AUTHORIZATION,
        "Content-Type": "application/json",
        "Accept": "application/json, text/event-stream",
    }
    if session_id:
        headers["mcp-session-id"] = session_id
    req = urllib.request.Request(
        REMOTE_URL,
        data=json.dumps(payload).encode("utf-8"),
        method="POST",
        headers=headers,
    )
    with urllib.request.urlopen(req, timeout=45) as resp:
        body = resp.read().decode("utf-8", "replace")
        new_session_id = resp.headers.get("mcp-session-id")
        if new_session_id:
            global SESSION_ID
            SESSION_ID = new_session_id
        content_type = resp.headers.get("content-type") or ""
        if "text/event-stream" in content_type:
            return parse_sse_body(body)
        if not body.strip():
            return None
        return json.loads(body)


def rewrite_cli_payload(payload: Dict[str, Any]) -> Dict[str, Any]:
    method = payload.get("method")
    if method != "tools/call":
        return payload
    params = payload.setdefault("params", {})
    tool_name = params.get("name")
    arguments = params.setdefault("arguments", {})
    if tool_name in {"companyBaseInfo", "risk"} and "companyName" in arguments and "keyword" not in arguments:
        arguments["keyword"] = arguments["companyName"]
    return payload


def handle_initialize(message: Dict[str, Any]) -> None:
    global SESSION_ID
    payload = {
        "jsonrpc": "2.0",
        "id": message["id"],
        "method": "initialize",
        "params": message.get("params", {}),
    }
    response = remote_post(payload, None)
    if response is None:
        write_message(
            {
                "jsonrpc": "2.0",
                "id": message["id"],
                "error": {"code": -32000, "message": "Tianyancha initialize returned no response"},
            }
        )
        return
    if "result" in response:
        capabilities = response["result"].setdefault("capabilities", {})
        capabilities.setdefault("resources", {"listChanged": False})
    write_message(response)


def handle_initialized() -> None:
    global INITIALIZED
    INITIALIZED = True
    try:
        remote_post({"jsonrpc": "2.0", "method": "notifications/initialized", "params": {}}, SESSION_ID)
    except Exception:
        return


def handle_forward(message: Dict[str, Any]) -> None:
    try:
        response = remote_post(rewrite_cli_payload(message), SESSION_ID)
        if response is not None:
            write_message(response)
        elif "id" in message:
            write_message({"jsonrpc": "2.0", "id": message["id"], "result": {}})
    except Exception as exc:
        if "id" in message:
            write_message(
                {
                    "jsonrpc": "2.0",
                    "id": message["id"],
                    "error": {"code": -32000, "message": f"Tianyancha bridge error: {exc}"},
                }
            )


def run_cli(tool_name: str, arguments: Dict[str, Any]) -> int:
    payload = rewrite_cli_payload(
        {
            "jsonrpc": "2.0",
            "id": "cli",
            "method": "tools/call",
            "params": {"name": tool_name, "arguments": arguments},
        }
    )
    try:
        response = remote_post({"jsonrpc": "2.0", "id": 1, "method": "initialize", "params": {"protocolVersion": "2024-11-05", "capabilities": {}, "clientInfo": {"name": "lead-discovery-cli", "version": "1.0.0"}}}, None)
        if response is None:
            print(json.dumps({"ok": False, "error": "No initialize response"}, ensure_ascii=False))
            return 1
        result = remote_post(payload, SESSION_ID)
        print(json.dumps(result, ensure_ascii=False))
        return 0
    except Exception as exc:
        print(json.dumps({"ok": False, "error": str(exc)}, ensure_ascii=False))
        return 1


def main() -> None:
    if len(sys.argv) >= 3:
        tool_name = sys.argv[1]
        try:
            arguments = json.loads(sys.argv[2])
        except Exception as exc:
            print(json.dumps({"ok": False, "error": f"Invalid JSON args: {exc}"}, ensure_ascii=False))
            raise SystemExit(1)
        raise SystemExit(run_cli(tool_name, arguments))

    while True:
        message = read_message()
        if message is None:
            break
        method = message.get("method")
        if method == "initialize":
            handle_initialize(message)
        elif method == "notifications/initialized":
            handle_initialized()
        elif method == "resources/list":
            write_message({"jsonrpc": "2.0", "id": message["id"], "result": {"resources": []}})
        elif method == "resources/templates/list":
            write_message({"jsonrpc": "2.0", "id": message["id"], "result": {"resourceTemplates": []}})
        else:
            handle_forward(message)


if __name__ == "__main__":
    main()
